dot/types/grandpa.go

```diff
@@ -196,3 +196,62 @@ type GrandpaVote struct {
 func (v GrandpaVote) String() string {
 	return fmt.Sprintf("hash=%s number=%d", v.Hash, v.Number)
 }
+
+// GrandpaEquivocation is used to create a proof of equivocation
+type GrandpaEquivocation struct {
+	RoundNumber     uint64
+	ID              [32]byte
+	FirstVote       GrandpaVote
+	FirstSignature  [64]byte
+	SecondVote      GrandpaVote
+	SecondSignature [64]byte
+}
+
+// GrandpaEquivocationEnum is a wrapper object for GRANDPA equivocation proofs, useful for unifying prevote
+// / and precommit equivocations under a common type.
+type GrandpaEquivocationEnum scale.VaryingDataType
+
+// Set sets a VaryingDataTypeValue using the underlying VaryingDataType
+func (ge *GrandpaEquivocationEnum) Set(value scale.VaryingDataTypeValue) (err error) {
+	vdt := scale.VaryingDataType(*ge)
+	err = vdt.Set(value)
+	if err != nil {
+		return err
+	}
+	*ge = GrandpaEquivocationEnum(vdt)
+	return nil
+}
+
+// Value will return the value from the underlying VaryingDataType
+func (ge *GrandpaEquivocationEnum) Value() (value scale.VaryingDataTypeValue, err error) {
+	vdt := scale.VaryingDataType(*ge)
+	return vdt.Value()
+}
+
+// NewGrandpaEquivocation returns a new VaryingDataType to represent a grandpa Equivocation
+func NewGrandpaEquivocation() *GrandpaEquivocationEnum {
+	vdt := scale.MustNewVaryingDataType(PreVoteEquivocation{}, PreCommitEquivocation{})
+	ge := GrandpaEquivocationEnum(vdt)
+	return &ge
+}
+
+// PreVoteEquivocation equivocation type for a prevote
+type PreVoteEquivocation GrandpaEquivocation
+
+// Index returns VDT index
+func (PreVoteEquivocation) Index() uint { return 0 }
+
+// PreCommitEquivocation equivocation type for a precommit
+type PreCommitEquivocation GrandpaEquivocation
+
+// Index returns VDT index
+func (PreCommitEquivocation) Index() uint { return 1 }
+
+// GrandpaOpaqueKeyOwnershipProof contains a key ownership proof for reporting equivocations
+type GrandpaOpaqueKeyOwnershipProof []byte
+
+// GrandpaEquivocationProof is used to report grandpa equivocations
+type GrandpaEquivocationProof struct {
```

Comment:
I think it may make sense to create a separate sub package to house these primitives that map 1 to 1 to the `primitives/finality-grandpa` crate.  Would need to do the same for BABE.

Reply:
I think this is related to my comment here, https://github.com/ChainSafe/gossamer/pull/3007#discussion_r1093771317

Let me know what you think